hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/MountVolumeMap.java

```diff
@@ -0,0 +1,143 @@
+
+import java.nio.channels.ClosedChannelException;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * MountVolumeMap contains information of the relationship
+ * between underlying filesystem mount and datanode volumes.
+ *
+ * This is useful when configuring block tiering on same disk mount (HDFS-15548)
+ * For now,
+ * we don't configure multiple volumes with same storage type on a mount.
+ */
+@InterfaceAudience.Private
+class MountVolumeMap {
+  private ConcurrentMap<String, Map<StorageType, VolumeInfo>>
+      mountVolumeMapping;
+  private double reservedForArchive;
+
+  MountVolumeMap(Configuration conf) {
+    mountVolumeMapping = new ConcurrentHashMap<>();
+    reservedForArchive = conf.getDouble(
+        DFSConfigKeys.DFS_DATANODE_RESERVE_FOR_ARCHIVE_PERCENTAGE,
+        DFSConfigKeys.DFS_DATANODE_RESERVE_FOR_ARCHIVE_PERCENTAGE_DEFAULT);
+    if (reservedForArchive > 1) {
+      FsDatasetImpl.LOG.warn("Value of reserve-for-archival is > 100%." +
+          " Setting it to 100%.");
+      reservedForArchive = 1;
+    }
+  }
+
+  FsVolumeReference getVolumeRefByMountAndStorageType(String mount,
+      StorageType storageType) {
+    if (mountVolumeMapping != null
+        && mountVolumeMapping.containsKey(mount)) {
+      try {
+        VolumeInfo volumeInfo = mountVolumeMapping
+            .get(mount).getOrDefault(storageType, null);
+        if (volumeInfo != null) {
+          return volumeInfo.getFsVolume().obtainReference();
+        }
+      } catch (ClosedChannelException e) {
+        FsDatasetImpl.LOG.warn("Volume closed when getting volume" +
+            " by mount and storage type: "
+            + mount + ", " + storageType);
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Return configured capacity ratio. Otherwise return 1 as default
+   */
+  double getCapacityRatioByMountAndStorageType(String mount,
+      StorageType storageType) {
+    if (mountVolumeMapping != null
+        && mountVolumeMapping.containsKey(mount)) {
+      return mountVolumeMapping
+          .get(mount).getOrDefault(storageType, null).getCapacityRatio();
+    }
+    return 1;
+  }
+
+  void addVolume(FsVolumeImpl volume) {
+    String mount = volume.getMount();
+    if (!mount.isEmpty()) {
+      Map<StorageType, VolumeInfo> storageTypeMap =
+          mountVolumeMapping
+              .getOrDefault(mount, new ConcurrentHashMap<>());
+      if (storageTypeMap.containsKey(volume.getStorageType())) {
+        FsDatasetImpl.LOG.error("Found storage type already exist." +
+            " Skipping for now. Please check disk configuration");
+      } else {
+        VolumeInfo volumeInfo = new VolumeInfo(volume, 1);
+        if (volume.getStorageType() == StorageType.ARCHIVE) {
+          volumeInfo.setCapacityRatio(reservedForArchive);
+        } else if (volume.getStorageType() == StorageType.DISK) {
+          volumeInfo.setCapacityRatio(1 - reservedForArchive);
```

Comment:
what if we have a mount with one single volume? Following the current implementation we may assign an unnecessary capacity ratio to it. We only need to calculate and assign the ratio for volumes sharing the same mount with others.

Reply:
Thats a good point. I will make the change to ignore the capacity ratio of the volume if there is only one on the mount.